### PR TITLE
[Task 1847360]2. add mysql firewall module 

### PR DESCRIPTION
**Pull request by andxu**
[AB#1847360](https://dev.azure.com/mseng/a4d27ce2-a42d-4b71-8eef-78cee9a9728e/_workitems/edit/1847360)

### CODE REVIEW

File: azure-toolkit-libs/azure-toolkit-mysql-lib/src/main/java/com/microsoft/azure/toolkit/lib/mysql/service/MySqlFirewallRule.java [hunk 1]
```diff
@@ -5,5 +5,41 @@
...
+        manager.firewallRules().delete(mySqlEntity.getResourceGroup(), this.mySqlEntity.getName(), this.entity.getName());
+    }
+
+    public ICommittable<MySqlFirewallRule> update(String startIpAddress, String endIpAddress) {
+        return () -> {
+            FirewallRule rule = manager.firewallRules().get(mySqlEntity.getResourceGroup(),
+                MySqlFirewallRule.this.mySqlEntity.getName(), MySqlFirewallRule.this.entity.getName()).update()
+                .withStartIpAddress(startIpAddress)
+                .withEndIpAddress(endIpAddress).apply();
+            MySqlFirewallRule.this.entity = fromFirewallRule(rule);
```
Comment (shenqianjin) [inline azure-toolkit-libs/azure-toolkit-mysql-lib/src/main/java/com/microsoft/azure/toolkit/lib/mysql/service/MySqlFirewallRule.java:41]: need we add a check existence logic to update?

File: azure-toolkit-libs/azure-toolkit-mysql-lib/src/main/java/com/microsoft/azure/toolkit/lib/mysql/service/MySqlFirewallRules.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        if (myFirewallRule != null) {
+            if (myFirewallRule.getEntity() != null && StringUtils.equals(myFirewallRule.getEntity().getStartIpAddress(), publicIp)) {
+                return myFirewallRule;
+            }
+            return myFirewallRule.update(publicIp, publicIp).commit();
+        } else {
+            return create().withName(name)
+                .wihStartIpAddress(IP_ALLOW_ACCESS_TO_AZURE_SERVICES)
+                .withEndIpAddress(IP_ALLOW_ACCESS_TO_AZURE_SERVICES).commit();
+        }
```
Comment (shenqianjin) [inline azure-toolkit-libs/azure-toolkit-mysql-lib/src/main/java/com/microsoft/azure/toolkit/lib/mysql/service/MySqlFirewallRules.java:56]: with wrong IP constant. -> publicIp.

File: azure-toolkit-libs/azure-toolkit-mysql-lib/src/main/java/com/microsoft/azure/toolkit/lib/mysql/service/MySqlFirewallRules.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        Optional.ofNullable(getAzureAccessRule()).ifPresent(MySqlFirewallRule::delete);
+    }
+
+    public void disableLocalMachineAccessRule() {
+        Optional.ofNullable(getAzureAccessRuleByName(getAccessFromLocalRuleName())).ifPresent(MySqlFirewallRule::delete);
+    }
+
+    private MySqlFirewallRule getAzureAccessRule() {
+        return getAzureAccessRuleByName(NAME_ALLOW_ACCESS_TO_AZURE_SERVICES);
+    }
```
Comment (shenqianjin) [inline azure-toolkit-libs/azure-toolkit-mysql-lib/src/main/java/com/microsoft/azure/toolkit/lib/mysql/service/MySqlFirewallRules.java:79]: how about to use "getAzureAccessRuleByName(NAME_ALLOW_ACCESS_TO_AZURE_SERVICES)" directly?

Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-mysql-lib/src/main/java/com/microsoft/azure/toolkit/lib/mysql/service/MySqlFirewallRules.java:79]: will refactor it in later PR

File: azure-toolkit-libs/azure-toolkit-mysql-lib/src/main/java/com/microsoft/azure/toolkit/lib/mysql/service/MySqlFirewallRules.java (hunk 2 above)
Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-mysql-lib/src/main/java/com/microsoft/azure/toolkit/lib/mysql/service/MySqlFirewallRules.java:56]: changed

File: azure-toolkit-libs/azure-toolkit-mysql-lib/src/main/java/com/microsoft/azure/toolkit/lib/mysql/service/MySqlFirewallRule.java (hunk 1 above)
Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-mysql-lib/src/main/java/com/microsoft/azure/toolkit/lib/mysql/service/MySqlFirewallRule.java:41]: this this class, user has already got an rule, so the rule shall be there, needn't to check existence

**Review by shenqianjin — APPROVED** [review]

**Pull request merged**